src/test/fuzz/hex.cpp

```diff
@@ -34,8 +32,14 @@ FUZZ_TARGET(hex)
         assert(Wtxid::FromHex(random_hex_string));
         assert(uint256::FromUserHex(random_hex_string));
     }
-    if (const auto result{uint256::FromUserHex(random_hex_string)}) {
-        assert(uint256::FromHex(result->ToString()));
+    if (auto result{uint256::FromUserHex(random_hex_string)}) {
+        auto result_string = result->ToString();
+        assert(result_string.length() == 64);
+        assert(IsHex(result_string));
+        assert(TryParseHex(result_string));
+        assert(Txid::FromHex(result_string));
+        assert(Wtxid::FromHex(result_string));
+        assert(uint256::FromHex(result_string));
```

Comment:
I'm not sure these 3 lines are worth adding, since they're already covered [higher up](https://github.com/bitcoin/bitcoin/pull/30618/files#diff-12dd31b274739b6a132d9c57e5ab14b403436991e4201cf8d3f3cc4fc7995066R29-R34), so this might unnecessarily slow down the test?

Reply:
those were testing the same input, this one is testing the parsing of `result->ToString()`

Reply:
Isn't that just a very expensive way to do `assert(random_hex_string == ToLower(result->ToString())`? (and also, that's not `hex` testing, but `uint256` testing)

Reply:
isn't the point of these tests that we shouldn't have know the internals (i.e. black box)?
(though in this case that wouldn't even be true since `random_hex_string` can start with `0x` for `FromUserHex` which isn't added back via `ToString`)

Reply:
> (though in this case that wouldn't even be true since `random_hex_string` can start with `0x` for `FromUserHex` which isn't added back via `ToString`)

Ah yes good point thanks, perhaps that would be good to clarify in a docstring?

> isn't the point of these tests that we shouldn't have know the internals (i.e. black box)?

I'm out of my depth here, so will probably refrain from commenting further and hopefully someone with more fuzzing experience can chime in. I think the point of the fuzz tests is to find bugs, and my understanding is that we often tweak them (e.g. based on knowledge about the implementation) to make them more effective at doing that. As such, not testing the `{Txid,Wtxid,uint256}::FromHex()` asserts twice seems like a reasonable improvement, but I'm not sure.

Reply:
On my machine, running

`FUZZ=hex build_fuzz/src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/hex -runs=1000000`

gives 681 seconds with those three lines and 533 seconds without them. Exec/s are reasonable (>1000) for both versions of the harness. The slow down isn't too drastic, so if these added checks are useful in any way then I think it's fine to keep them.

Reply:
Thanks for looking into it @marcofleon! If there are no operational concerns adding the test cases, I'm happy with it.

I guess if we keep this, the `uint256::FromHex()` asserts have become superfluous and could be removed. Shouldn't make a meaningful performance impact (since they're called much less often than the `FromUserHex()` asserts, so it's more of a thing to keep the test code lean.

<details>
<summary>git diff on 6fa4c92e4c</summary>

```diff
diff --git a/src/test/fuzz/hex.cpp b/src/test/fuzz/hex.cpp
index 1109186a8f..71518f626a 100644
--- a/src/test/fuzz/hex.cpp
+++ b/src/test/fuzz/hex.cpp
@@ -28,12 +28,7 @@ FUZZ_TARGET(hex)
     if (IsHex(random_hex_string)) {
         assert(ToLower(random_hex_string) == hex_data);
     }
-    if (uint256::FromHex(random_hex_string)) {
-        assert(random_hex_string.length() == 64);
-        assert(Txid::FromHex(random_hex_string));
-        assert(Wtxid::FromHex(random_hex_string));
-        assert(uint256::FromUserHex(random_hex_string));
-    }
+    (void)uint256::FromHex(random_hex_string);
     if (auto result{uint256::FromUserHex(random_hex_string)}) {
         const auto result_string{result->ToString()};
         assert(result_string.length() == 64);

```
</details>

Reply:
I don’t fully understand why we’re focusing on optimization at this stage. Why not keep both options for now, and if we later find that performance is an issue, remove the redundant ones based on their actual implementations? In my opinion, we’re still in the discovery phase, where having as much redundancy as we can afford is beneficial. Since fuzzing—and property-based testing in general—is black-box testing, we acknowledge that we don’t know the full input space. Our goal is to ensure that certain properties are upheld, even though we might not fully understand how those properties could be violated.

Reply:
> We seem to disagree

We don't have to, if the above discussion didn't change your mind I'll gladly remove them.

Reply:
Done

Reply:
I'm hesitant to add more to the back-and-forth, but I think we may have been talking past each other. In my [latest comment](https://github.com/bitcoin/bitcoin/pull/30618#discussion_r1744124233) I suggested updating the `FromHex()` tests, where based on the [latest force-push](https://github.com/bitcoin/bitcoin/compare/6fa4c92e4c9bf5ea90e25d41b525c88f9c2d834d..2cee181a6cc00840def16c311aae69883130eebd) it seems you were talking about the `FromUserHex()` tests, which I agreed I was okay with based on @marcofleon's [testing](https://github.com/bitcoin/bitcoin/pull/30618#discussion_r1743946931).

@marcofleon  also showed me (and I agree with) that removing `assert(uint256::FromHex(result_string));` is a slight regression since it's not a duplicate (and [I specifically added it](https://github.com/bitcoin/bitcoin/pull/30569/commits/70e2c87737e77ee85812cc328c4ddfaea7147533#diff-12dd31b274739b6a132d9c57e5ab14b403436991e4201cf8d3f3cc4fc7995066R38)), so that could be added back in? Sorry for not highlighting that earlier, I thought we'd moved past keeping all the `FromUserHex()` tests already so I didn't want to add noise.

Reply:
> I'm hesitant to add more to the back-and-forth

I don't mind jousting until we find something that is better than what either of us had in mind. Happens often, keep insisting on parts that you feel strongly about!
Let's see what you think of my latest push:
* FromHex correctness assumes FromUserHex parseability
* FromUserHex correctness assumes that if reconverted to hex, it should be accepted by FromHex

Pushed (first a separate rebase, and a change)

Reply:
> Let's see what you think of my latest push:

Yeah I like the current state of it, thanks for addressing this and can be marked as resolved now!